### ISSUE TITLE
Kubernetes Custom Resource Scaler

### ISSUE THREAD

**Issue by kate-goldenring** [kedacore/keda#2251]
### Proposal

KEDA has a [Kubernetes workload](https://keda.sh/docs/2.4/scalers/kubernetes-workload/) scaler, but it would be great if there was a scaler with Kubernetes Custom Resources as the trigger. Say if you wanted the creation of a custom resource to trigger the deployment of jobs. This would enable KEDA acting almost as a Kubernetes operator for CRDs.

### Scaler Source

Any custom resource. Maybe you could specify the resource name.

### Scaling Mechanics

Upon the creation of custom resources, KEDA could scale up Pods, Services, other Custom Resources, or a combination of these.

### Authentication Source

same auth scenario as the  [Kubernetes workload](https://keda.sh/docs/2.4/scalers/kubernetes-workload/) scaler should work

**Comment by JorTurFer** [comment]
Hi!
sorry, maybe I'm missing anything, but... where is the improvement if you have to apply a manifest to spawn a job? I mean, you could apply the job manifest instead of another CRD, couldn't?

**Comment by kate-goldenring** [comment]
> Hi! sorry, maybe I'm missing anything, but... where is the improvement if you have to apply a manifest to spawn a job? I mean, you could apply the job manifest instead of another CRD, couldn't?

I am thinking of scenarios where the purpose of the CRD isnt just to kick off job creation -- I agree that that does seem a bit pointless. 

I'm thinking of scenarios where custom resources are dynamically created and removed. Having associated resources (namely pods) be created and removed would be ideal. 

This may be niche and something others are not interested, but I'll give some context on where I'm coming from. I maintain [Akri](https://github.com/project-akri/akri) which discovers iot devices around kubernetes clusters on the edge and automatically deploys workloads to use the devices. We create a custom resource (Instance) to represent each device. We currently have our own controller that deploys pods upon the creation of the Instances; however, i was curious if we could find a way to use KEDA instead. This could potentially enable users to choose their own scaling strategies; rather than our controller supporting them all.

**Comment by arschles** [comment]
I agree - letting users scale from CRDs _unrelated to KEDA_ would provide a good connection point to other systems.

@kate-goldenring what do you think the `trigger` section should look like?

**Comment by JorTurFer** [comment]
> > Hi! sorry, maybe I'm missing anything, but... where is the improvement if you have to apply a manifest to spawn a job? I mean, you could apply the job manifest instead of another CRD, couldn't?
> 
> I am thinking of scenarios where the purpose of the CRD isnt just to kick off job creation -- I agree that that does seem a bit pointless.
> 
> I'm thinking of scenarios where custom resources are dynamically created and removed. Having associated resources (namely pods) be created and removed would be ideal.
> 
> This may be niche and something others are not interested, but I'll give some context on where I'm coming from. I maintain [Akri](https://github.com/project-akri/akri) which discovers iot devices around kubernetes clusters on the edge and automatically deploys workloads to use the devices. We create a custom resource (Instance) to represent each device. We currently have our own controller that deploys pods upon the creation of the Instances; however, i was curious if we could find a way to use KEDA instead. This could potentially enable users to choose their own scaling strategies; rather than our controller supporting them all.

Oh! I got it!
It makes sense :)
I stay tuned to your vision of the trigger

**Comment by JorTurFer** [comment]
I'm thinking on this and I'm not sure about if this is doable. I mean, KEDA should work in clusters with and without RBAC. In clusters without RBAC, we can get all the resources that we want, but if the RBAC is enabled, I'm not sure about how we can query the resources.
My concert here is the required permissions for KEDA, we should list all resources all namespaces and I don't know if it's worth having that privilege

**Comment by tomkerkhove** [comment]
> My concert here is the required permissions for KEDA, we should list all resources in all namespaces and I don't know if it's worth having that privilege. IDK if we can grant permissions generically over APIs which name we don't know

If it doesn't have the permissions it will not see them and move on, so it's up to the cluster operator to manage that IMO, no?

I think this is a nice use-case, but probably not an easy one given we'd know what CRDs to look for and what to scale so echoing @arschles his question on what it would look like.

Also, that means that our operator will have to monitor Kubernetes resources for that CRD and I'm not sure we can easily do that. If that is the case, then your operator (in this case Akri) will have to send us notifications or events that it was added/deleted. CloudEvents should be used here for these notifications.

I like the idea but am curious about how we will implement this.

What are your thoughts on this @zroubalik?

**Comment by JorTurFer** [comment]
I'm still thinking on this...
Could this be a case for an external scaler? I mean, Akri (or other operator) knows its own CRDs and has enough permissions to manage them. 
We already can use those values if another service wrappes them for us (external scaler, metrics api scaler) 

For standard k8s apis I can imagin how to do it, but for CRDs...

**Comment by tomkerkhove** [comment]
External scaler is certainly one way, but I see a bigger fit for this than Akri.

**Comment by JorTurFer** [comment]
That's truth, I only wanted to give another extra option to evaluate if this is not doable :)

**Comment by zroubalik** [comment]
I would like to see what the trigger section could look like as @arschles mentioned. 

I am not 100% about this one, but might not understand the request properly. But the usecase for me looks like that we are trying to fit KEDA where the custom controller should be implemented. I mean, that's the Operator pattern - watch particular resources and do some stuff (ie deploy pods, services) once I create the CR. But I might be wrong, so looking forward to see the trigger spec draft.

**Comment by tomkerkhove** [comment]
It feels like an extension to me of the operator but I do get your point - What I understood was that when a resource is created, the operator will do its magic and in parallel we want to scale another workloads.

**Comment by kate-goldenring** [comment]
I havent looked deeply into the KEDA implementation but from my understanding it would be the scaler that would need to do the querying. Could steps be added to install instructions to make sure they have a role and rolebinding for the scaler to access the resource?

**Comment by kate-goldenring** [comment]
> I agree - letting users scale from CRDs _unrelated to KEDA_ would provide a good connection point to other systems.
> 
> @kate-goldenring what do you think the `trigger` section should look like?

I could see it being similar to the kubernetes-workload one -- i feel like much of that implementation could probably be reused if the trigger was:
```
triggers:
- type: custom-resource-workload
  metadata:
    kindSelector: 'name=Instance'
    value: '1'
```
where `kindSelector` specifies the name of the CRD (Instance in this case) and value like the k8s workload specifies the number of pods (or jobs ideally) that would be deployed for each one.

**Comment by kate-goldenring** [comment]
> I would like to see what the trigger section could look like as @arschles mentioned.
> 
> I am not 100% about this one, but might not understand the request properly. But the usecase for me looks like that we are trying to fit KEDA where the custom controller should be implemented. I mean, that's the Operator pattern - watch particular resources and do some stuff (ie deploy pods, services) once I create the CR. But I might be wrong, so looking forward to see the trigger spec draft.

This is a fair point -- this is putting KEDA in the role of monitoring resources and being an operator. So it might not be the right fit -- or it could be expanding KEDA to a new area. I imagine a simple prepackaged operator that is a one use case operator: see a resource and scale.

**Comment by arschles** [comment]
>this is putting KEDA in the role of monitoring resources and being an operator

that's indeed true. I do think that this would be generally useful functionality because one could use it to tie KEDA into a vast array of other CRD-based systems.

@kate-goldenring is the intention of that trigger to just scale based on the _existence_ of an arbitrary CRD? I personally don't think that's out of scope of KEDA, assuming the project is willing to take on the documentation and (possibly) troubleshooting burden of educating folks about how this feature interacts with RBAC.

> that means that our operator will have to monitor Kubernetes resources for that CRD and I'm not sure we can easily do that

I think we can, using the [dynamic client](https://pkg.go.dev/k8s.io/client-go@v0.22.3/dynamic). This client gives you access to the [`ResourceInterface`](https://pkg.go.dev/k8s.io/client-go@v0.22.3/dynamic#ResourceInterface) type, which you can then use to get an [`Unstructured`](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured#Unstructured). `Unstructured` is a simple wrapper around a `map[string]interface{}`, which we can't generically work with. I don't think we'll have to, though. As long as we're not doing scaling based on the _contents_ of a CRD and just doing it based on the _existence_ of one, the code for this scaler might look like the below:

```go
dynIface, err := dynamic.NewForConfig(restConf)
if err != nil {
    // ...
    panic(err)
}
crdGVK := schema.GroupVersionResource {
    // These three fields arbitrary, and in reality would be taken from the ScaledObject trigger
    Group: "my.crd.group",
    Version: "v1alpha1", 
    Name: "HTTPScaledObject",
}
nsResourceIface := dynIface.Resource(crdGVK)
obj, err := nsResourceIface.Get(ctx, "something", metav1.GetOptions{})
// you can check for IsNotFound, etc... here
if err != nil {
    panic(err)
}
// we probably don't have much to do with obj here, since we're just checking for existence
fmt.Println(obj)
```

**Comment by tomkerkhove** [comment]
Agreed. If we don't want to make this part of the core, this could also be an add-on similar to HTTP add-on that we provide but don't ship in the core due to the security burden.

Thoughts @zroubalik ?

**Comment by zroubalik** [comment]
@tomkerkhove starting as a add-on is an approach I'd vote for.

**Comment by arschles** [comment]
@kate-goldenring and I spoke on a video call yesterday about this. It looks like an external scaler (i.e. add-on) will work fine for this. I believe (@kate-goldenring , please correct me if I'm wrong) she is going to do some more investigation on her side to figure out the best way to make Akri and KEDA play well together

**Comment by kate-goldenring** [comment]
From @arschles and I's discussion, i realized that while KEDA changes the # of replicas of deployments, it cannot modify the deployment itself (namely the `ScaledObjectRef`). With Akri, we'd want the pod to be modified (namely a resource request to be added) depending on the Instance (our CRD). This isnt a case that KEDA can cover currently. It looks like KEDA could scale deployments created by Akri Controller but could not be a replacement. Furthermore, Akri currently only creates Pods not deployments or jobs as supported by KEDA. If this changes, there may be a change to integrate. I will reconsider whether an integration will be advantageous if/when we support non-Pod deployments in our controller. Thank you for all the thoughts here!

**Comment by tomkerkhove** [comment]
Feel free to circle back with how you have fixed it, if you can.Who knows there is some way we can help

**Comment by JorTurFer** [comment]
I closed this by error

**Comment by JorTurFer** [comment]
I guess that this solves the case https://keda.sh/docs/2.19/scalers/kubernetes-resource/
